### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by RWOverdijk** [inkle/ink#619]
I'm just now getting into ink and plan on using it with unity.  I'm at "Part 3: Variables and Logic" in the "Writing with ink" right now, and I am beyond thoroughly impressed so far.

But also I am curious. Given all these features, doesn't this create a lot of overhead performance wise? I mean, parsing the json and all is one thing, but keeping track of all these choices and variables... It just seems like it would be very heavy.

Have you noticed any issues with this? I get that convenience always comes with a cost in terms of performance, I am just curious what the cost is with ink as it looks too good to be true.

It's not going to stop me from having fun with this. 😄

**Comment by joethephish** [comment]
It's all very relative to exactly how you use ink!

Most stories aren't massively complex, even if they're large - for example massively branching stories in the style of [80 Days](https://www.inklestudios.com), so the ink processing time is pretty minimal, even it has hundreds of thousands of words (story JSON data is around 10MB I think).

On the other end of the spectrum, if you decide to use ink as a general purpose scripting language, it won't be ideal - using it for complex calculations with lots of recursion etc will definitely get slow. That said, we do some pretty complex logic stuff in [Heaven's Vault](https://www.inklestudios.com/heavensvault) and [Pendragon](https://www.inklestudios.com/pendragon), and although it can cause a bit of processing time in the engine, you can evaluate asynchronously, and you can also optimise by turning ink functions into external functions that are run in C# (or JavaScript if that's what you're using).

**Comment by RWOverdijk** [comment]
Cool :) I'm definitely going to keep most of the hard work in C# I think, seems most logical. But it's really nice to know I can do smaller repetitive things in ink directly.

I know this issue is about performance, so feel free to refuse to answer the following 😄

With unity I want to have an open world and trigger conversations as you approach NPCs. I saw `ChoosePathString` and figured that makes sense, but it leaves one question.

Does that mean that after a knot has knit (lol) it doesn't _end_? Because in the inky player it does. So I'm not sure how to keep it running and call parts myself.

Basically it'll be a main storyline triggered by events, and subquests on the side. I'm curious how to structure that if it ends at any point.

**Comment by joethephish** [comment]
A good way to get around this is to turn everything into a choice, and then you can test more easily in Inky. Eg a choice like `+ [Bob - talk]`. That allows the ink to dictate to the decision of who you’re allowed to talk to, or what you’re able to interact with. (But obviously on the game side you can also choose to not allow interactions with those characters.)

To make this scalable it helps to understand ink threads, so that you can pull in blocks of reusable choices (that may correspond to common in-game actions) from various places.
